packages/hooks/src/data/OperationData.ts

```diff
@@ -0,0 +1,114 @@
+import { ApolloClient } from 'apollo-client';
+import { isEqual } from 'apollo-utilities';
+import { invariant } from 'ts-invariant';
+import {
+  ApolloContextValue,
+  parser,
+  DocumentType,
+  operationName
+} from '@apollo/react-common';
+import { DocumentNode } from 'graphql';
+
+import { CommonOptions } from '../types';
+
+export abstract class OperationData<TOptions = any> {
+  private _isMounted: boolean = true;
+  private _options: CommonOptions<TOptions> = {} as CommonOptions<TOptions>;
+  private _previousOptions: CommonOptions<TOptions> = {} as CommonOptions<
+    TOptions
+  >;
+  private _context: ApolloContextValue = {};
+  private _client: ApolloClient<object> | undefined;
+
+  constructor(options?: CommonOptions<TOptions>, context?: ApolloContextValue) {
+    this._options = options || ({} as CommonOptions<TOptions>);
+    this._context = context || {};
+  }
+
+  get isMounted(): boolean {
+    return this._isMounted;
+  }
+
+  set isMounted(mounted: boolean) {
+    this._isMounted = mounted;
+  }
+
+  get options(): CommonOptions<TOptions> {
+    return this._options;
+  }
+
+  set options(newOptions: CommonOptions<TOptions>) {
+    if (!isEqual(this.options, newOptions)) {
+      this.previousOptions = this.options;
```

Comment:
Even this logic might be better implemented by an explicit `getOptions`/`setOptions` method pair, rather than a getter and a setter.